File: Java/com/agodi/QuickSort.java
package com.agodi;

import java.util.Arrays;

public class QuickSort {

	public static void main(String[] args) {
		int[] array = { 0, 9, 8, 7, 6, 5, 4, 3, 2, 1 };
		sort(array, 0, array.length);
		System.out.println(Arrays.toString(array));
	}

	public static void sort(int[] array, int i, int j) {
        if(i==j) return;
		int pivotIndex = partition(array, i, j);
		sort(array, i, pivotIndex);
		sort(array, pivotIndex + 1, j);
	}

	private static int partition(int[] array, int i, int j) {
		System.out.println(Arrays.toString(array));
		int index = i;//new Random().nextInt(j-i) + i;
		int pivot = array[index];
        System.out.println(pivot);
		array[index] = array[i];
		index = i;
		array[i] = pivot;
        System.out.println(Arrays.toString(array));
        i++;
        j--;
		while(i < j) {
			while(i < array.length && array[i] < pivot) {
                System.out.println(array[i]);
                i++;}
			while(j >=0 && array[j] > pivot) {
                System.out.println(array[j]);
                j--;}
            if(i<j) {
                int aux = array[i];
                array[i] = array[j];
                array[j] = aux;
            }
			System.out.println(Arrays.toString(array));
		}
        i--;
		array[index] = array[i];
		array[i] = pivot;
		System.out.println(Arrays.toString(array));
		return i;
	}

}

File: Python/quicksort.py
import random

def quicksort(integers):
    """Sorts a given array of integers using
        quicksort algorithm"""
    sort(integers, 0, len(integers))
    print "final array: ", integers
    
def sort(integers, start, stop):
    """Recursive function to split array in 
        two parts"""
    if start != stop:
        index = find_pivot_index(integers, start, stop)
        sort(integers, start, index)
        sort(integers, index+1, stop)

def find_pivot_index(integers, start, stop):
    """Function to pick a random element from the array
        and place it in the right position rearranging
        all smaller elements to it's left and all
        higher elements to it's right"""
    index = None
    pivot = random.randrange(start, stop)
    switch(integers, start, pivot)
    i = start + 1
    j = stop - 1
    while i <= j:
        while i < stop and integers[i] < integers[start]:
            i += 1
        while j > start and integers[j] >= integers[start]:
            j -= 1
        if i < j:
            switch(integers, i, j)
    index = i-1
    switch(integers, start, index)
    return index
    
def switch(integers, i, j):
    """Swaps two elements in the array"""
    if i == j:
        return
    aux = integers[i]
    integers[i] = integers[j]
    integers[j] = aux
    
integers = [1,8,7,6,5,4,3,2,1]
quicksort(integers)